client/client.go

```diff
@@ -396,39 +253,132 @@ type client struct {
 	tokenDispatcher *tokenDispatcher
 
 	// For internal usage.
-	checkTSDeadlineCh    chan struct{}
-	leaderNetworkFailure int32
+	checkTSDeadlineCh         chan struct{}
+	checkTSODispatcherCh      chan struct{}
+	updateTSOConnectionCtxsCh chan struct{}
+	updateTokenConnectionCh   chan struct{}
+	leaderNetworkFailure      int32
+	wg                        sync.WaitGroup
+
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	option *option
+}
+
+// SecurityOption records options about tls
+type SecurityOption struct {
+	CAPath   string
+	CertPath string
+	KeyPath  string
+
+	SSLCABytes   []byte
+	SSLCertBytes []byte
+	SSLKEYBytes  []byte
 }
 
 // NewClient creates a PD client.
-func NewClient(pdAddrs []string, security SecurityOption, opts ...ClientOption) (Client, error) {
-	return NewClientWithContext(context.Background(), pdAddrs, security, opts...)
+func NewClient(svrAddrs []string, security SecurityOption, opts ...ClientOption) (Client, error) {
+	return NewClientWithContext(context.Background(), svrAddrs, security, opts...)
 }
 
 // NewClientWithContext creates a PD client with context.
-func NewClientWithContext(ctx context.Context, pdAddrs []string, security SecurityOption, opts ...ClientOption) (Client, error) {
-	log.Info("[pd] create pd client with endpoints", zap.Strings("pd-address", pdAddrs))
+func NewClientWithContext(ctx context.Context, svrAddrs []string, security SecurityOption, opts ...ClientOption) (Client, error) {
+	log.Info("[pd] create pd client with endpoints", zap.Strings("pd-address", svrAddrs))
+	c, clientCtx, clientCancel, tlsCfg := createClient(ctx, &security)
+	c.tsoStreamBuilderFactory = &pdTSOStreamBuilderFactory{}
+	c.bc = newPDBaseClient(clientCtx, clientCancel, &c.wg, addrsToUrls(svrAddrs), tlsCfg, c.option)
+	if err := c.setup(true, true, opts...); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// NewTSOClientWithContext creates a TSO client with context.
+func NewTSOClientWithContext(ctx context.Context, svrAddrs []string, security SecurityOption, opts ...ClientOption) (Client, error) {
+	log.Info("[pd(tso)] create tso client with endpoints", zap.Strings("tso-address", svrAddrs))
+	c, clientCtx, clientCancel, tlsCfg := createClient(ctx, &security)
+	c.tsoStreamBuilderFactory = &tsoTSOStreamBuilderFactory{}
+	c.bc = newTSOBaseClient(clientCtx, clientCancel, &c.wg, addrsToUrls(svrAddrs), tlsCfg, c.option)
+	if err := c.setup(true, false, opts...); err != nil {
```

Comment:
I prefer not to use options. The client shouldn't know anything about if the tso or resource manager is enabled.

Reply:
I didn't notice here previously.

Reply:
I removed the flags for now. The original purpose of adding these flags is to disable some clients when the server side doesn't provide the correspoinding services. For example, disable tso client when the server side is resource management microservice, or disable resource management client when the server side the tso microservice.